File: spatial_audio/spatial_audio.py
"""
Spatial audio coordinate math — pure Python port of spatial-audio.js.

No audio hardware is used here.  The class stores the computed 3-D
position and listener-orientation vectors so callers (and tests) can
read them back directly.  A real application would forward these values
to the browser over the bridge WebSocket so the browser's PannerNode
and AudioListener can use them.
"""

import math


class SpatialAudio:
    def __init__(self):
        self._is_attached = False
        # Source position (set by set_source_position)
        self.position_x = 0.0
        self.position_y = 0.0
        self.position_z = 0.0
        # Listener forward vector (set by set_listener_orientation)
        self.forward_x = 0.0
        self.forward_y = 0.0
        self.forward_z = -1.0  # default: facing forward (-Z)

    # ── Lifecycle ────────────────────────────────────────────────────────────

    @property
    def is_attached(self):
        return self._is_attached

    def attach(self):
        """Simulate attaching an audio track (replaces any existing attachment)."""
        if self._is_attached:
            self._reset_state()
        self._is_attached = True

    def detach(self):
        """Release the attachment and reset all state."""
        if not self._is_attached:
            return
        self._is_attached = False
        self._reset_state()

    # ── Coordinate math ──────────────────────────────────────────────────────

    def set_source_position(self, azimuth: float, distance: float):
        """
        Convert (azimuth°, distance) to Cartesian (x, y, z).

        azimuth=0   → source straight ahead  (z = -distance, x = 0)
        azimuth=90  → source to the right    (x = +distance, z ≈ 0)
        """
        if not self._is_attached:
            return
        rad = (azimuth * math.pi) / 180.0
        self.position_x = distance * math.sin(rad)
        self.position_y = 0.0
        self.position_z = -distance * math.cos(rad)

    def set_listener_orientation(self, azimuth: float):
        """
        Set the listener's forward vector from a yaw azimuth in degrees.

        azimuth=0   → forward (0, 0, -1)
        azimuth=90  → right   (1, 0,  0)
        """
        if not self._is_attached:
            return
        rad = (azimuth * math.pi) / 180.0
        self.forward_x = math.sin(rad)
        self.forward_y = 0.0
        self.forward_z = -math.cos(rad)

    # ── Internal ─────────────────────────────────────────────────────────────

    def _reset_state(self):
        self.position_x = 0.0
        self.position_y = 0.0
        self.position_z = 0.0
        self.forward_x = 0.0
        self.forward_y = 0.0
        self.forward_z = -1.0

File: client/spatial-audio.js
class SpatialAudio {
  constructor() {
    this._ctx = null;
    this._source = null;
    this._panner = null;
  }

  get isAttached() {
    return this._ctx !== null;
  }

  attachTrack(track) {
    if (this._ctx) this._teardown();

    const ctx = new AudioContext();
    this._ctx = ctx;

    const stream = new MediaStream([track]);
    const source = ctx.createMediaStreamSource(stream);
    this._source = source;

    this._connectPanner(ctx, source);
  }

  attachElement(videoEl) {
    if (this._ctx) this._teardown();

    const ctx = new AudioContext();
    this._ctx = ctx;

    const source = ctx.createMediaElementSource(videoEl);
    this._source = source;

    this._connectPanner(ctx, source);
  }

  _connectPanner(ctx, source) {
    const panner = ctx.createPanner();
    panner.panningModel = 'HRTF';
    panner.distanceModel = 'inverse';
    panner.refDistance = 1;
    panner.maxDistance = 10000;
    panner.rolloffFactor = 1;
    this._panner = panner;

    source.connect(panner);
    panner.connect(ctx.destination);
  }

  setSourcePosition(azimuth, distance) {
    if (!this._panner) return;

    const rad = (azimuth * Math.PI) / 180;
    this._panner.positionX.value = distance * Math.sin(rad);
    this._panner.positionY.value = 0;
    this._panner.positionZ.value = -distance * Math.cos(rad);
  }

  setListenerOrientation(azimuth) {
    if (!this._ctx) return;

    const rad = (azimuth * Math.PI) / 180;
    this._ctx.listener.forwardX.value = Math.sin(rad);
    this._ctx.listener.forwardY.value = 0;
    this._ctx.listener.forwardZ.value = -Math.cos(rad);
  }

  async detach() {
    if (!this._ctx) return;
    this._teardown();
  }

  _teardown() {
    if (this._source) {
      this._source.disconnect();
      this._source = null;
    }
    if (this._panner) {
      this._panner.disconnect();
      this._panner = null;
    }
    if (this._ctx) {
      this._ctx.close();
      this._ctx = null;
    }
  }
}

if (typeof module !== 'undefined' && module.exports) {
  module.exports = { SpatialAudio };
}
